doc/mgmt/SONiC_overlay_ConfigDB.md

```diff
@@ -0,0 +1,234 @@
+
+# 2 Design
+
+- Layered config DB design diagram:
+
+<img src="./images/swss-common-layer.png"  />
+
+- API design diagram:
+
+<img src="./images/swss-common-default-value.png"  />
+
+## 2.1 Considerations
+
+### How to get default value
+
+|                                                               | Pros                                              | Cons                                                                                                                                              |
+| ------------------------------------------------------------- | ------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Get default value from Yang model in read API.                | Redis config DB keeps no change.                  | 3 MB memory per-process because need load Yang model and reference libyang.<br>50ms to load yang model.<br>8ms to read default value 10000 times. |
+| Write default value to default value DB when write config DB. | Better read performance, Less memory consumption. | Need add new Redis DB for default value.                                                                                                          |
+
+### How to get static config
+
+- Static config will stored in a new redis database 'STATIC_CONFIG_DB'
+- Static config DB tables will have exactly same name and schema with config DB tables.
+- Data will read form static config DB with swsscommon API.
+- Static config DB will save and persist for warm-reboot and fast-reboot.
+
+### API compatibility
+
+|                                                       | Pros                                                                              | Cons                                                                                                                                     |
+| ----------------------------------------------------- | --------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
+| Change API to return default value and static config. | Less code change, all app will get default value and static config automatically. | For default value, there are hardcoded default value may different with Yang model, new default value from config DB may cause code bug. |
+| Existed API keeps no change.                          | When update existed code, can cleanup code to remove hard coded default value.    | All apps need code update.                                                                                                               |
+
+### Current design:
+
+- Existed read API keeps no change. 
+- Add decorator API to return default value and static config data.
+
+## 2.2 New class
+
+- YangModelLoader class
```

Comment:
>YangModelLoader class [](http://example.com/codeflow?start=2&length=21)

Is it a new class or just extension of existing ConfigMgmt class?

Reply:
It's a new C++ class, and it's a internal class, not visible to user.

Reply:
I realize it may be a re-implementation of ConfigMgmt  in C++. People will ask for a single implementation, and wrapper in various other programming languages.